### ISSUE TITLE
Flycheck won't use python3 flake8 no more

### ISSUE THREAD

**Issue by tsoernes** [flycheck/flycheck#1409]
After package update in spacemacs, flycheck defaults to flake8 for python2. No flycheck-python-xx-executable was set, nor do I use virtualenv or anaconda. 
'which flake8' ponts to '/usr/bin/flake8', which is python3:
```
>> which flake8
/usr/bin/flake8
>> flake8 --version
3.5.0 (mccabe: 0.6.1, pycodestyle: 2.3.1, pyflakes: 1.6.0) CPython 3.6.4 on Linux
>> which python
/usr/bin/python
>>  python --version
Python 2.7.14
>> which python3
/usr/bin/python3
>> python3 --version
Python 3.6.4
```
Flake8 at `usr/bin/flake8` is detectable by executable-find. 
I tried fixing it by doing `pip2 uninstall flake8`, but flycheck still is not able to find it.
Have tried `setq-default flycheck-python-flake8-executable` to `flake8` and `(executable-find flake8)`, no luck. This is current Flycheck status:

```
  python-flake8 (disabled)
    - may enable:         Automatically disabled!
    - executable:         Found at /usr/bin/python
    - configuration file: Not found
    - `flake8' module:    Missing

  python-pylint (disabled)
    - may enable:         Automatically disabled!
    - executable:         Found at /usr/bin/python
    - configuration file: Not found
    - `pylint' module:    Missing

  python-pycompile
    - may enable: yes
    - executable: Found at /usr/bin/python

Flycheck Mode is enabled. Use SPC u C-c ! x to enable disabled checkers.
```
Fedora 27, Spacemacs

**Comment by tsoernes** [comment]
Today's package update in Spacemacs made `(setq-default flycheck-python-flake8-executable "flake8")` work. Why it is no longer able to find default to the system-wide flake8 executable I don't know.

**Comment by cpitclaudel** [comment]
Sorry for the trouble.  I can explain: Flycheck now runs flake8 as python -m flake8.  Your system's python is python2, so Flycheck uses that and doesn't find flake8.  The fix is to set the executable to "python3".  Your fix also works because I of backwards compatibility wrappers that I added.  Does that make sense? :)

**Comment by tsoernes** [comment]
Sure. If I want to use python2 and corresponding flake8 later though, do I
have to switch flake8 path manually in the emacs dotfile?

2018-02-18 19:57 GMT+01:00 Clément Pit-Claudel <notifications@github.com>:

**Comment by cpitclaudel** [comment]
Yes, though this switch could be done automatically by packages like venv.  How did you switch between flake8 for python2 and flake8 for python3 before?

**Comment by tsoernes** [comment]
I didnt, I just turned it off. Anyway to have flycheck decide based on the
shebang at the top of the file eg "#! /usr/bin/python3" ?

**Comment by cpitclaudel** [comment]
> Anyway to have flycheck decide based on the shebang at the top of the file eg "#! /usr/bin/python3" ?

Not directly in Flycheck, but the change that you ran into is intended to make this easy.
https://www.reddit.com/r/emacs/comments/3anrqf/setting_up_emacs_to_support_both_python_2_and/ is relevant.

I use something like the following:

```
(save-match-data
  (save-excursion
    (goto-char (point-min))
    (when (looking-at "#![ \t]?\\(?:[^ \t\n]*/bin/env[ \t]\\)?\\([^ \t\n]+\\)")
      (let ((python (match-string-no-properties 1)))
        (setq-local python-shell-interpreter python)
        (setq-local flycheck-python-pylint-executable python)))))
```
